c2rust-analyze/src/main.rs

```diff
@@ -278,6 +278,35 @@ fn update_pointer_info<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>)
     }
 }
 
+// Walks the type `ty` and applies a function `f` to it if it's an ADT
+// `f` gets applied recursively to `ty`'s generic types and fields (if applicable)
+fn walk_args_and_fields<'tcx, F>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, f: &mut F)
+where
+    F: FnMut(DefId) -> bool,
+{
+    // the first type encountered in the walk is `ty`, and the
+    // relevant handling for that is below, so we can skip it here
+    for arg in ty.walk().skip(1) {
+        if let GenericArgKind::Type(ty) = arg.unpack() {
+            walk_args_and_fields(tcx, ty, f);
+        }
+    }
```

Comment:
I don't think we want to walk the generics.  A type could be a generic type parameter but not actually be in the type layout (e.x. `PhantomData`).

Reply:
This loop also walks array elements, pointees of raw pointers, and such, which we don't want to ignore.

Reply:
Given [this comment](https://github.com/immunant/c2rust/pull/960#discussion_r1231639983), marking resolved

Reply:
The way `walk` walks types is kind of misleading in the name. It implicitly casts types and consts like array lengths to `GenericArgs` for convenience. It still iterates over all reachable types regardless of whether or not the item is a generic arg. For `FnPtrs` it iterates over the inputs and outputs